### PR TITLE
Add maximum version to EndpointMinimumVersion definitions and rename where necessary

### PR DESCRIPTION
**Pull request by ylee088**
## Before this PR

Currently only the minimum versions for the endpoints are included, and we want to include maximum versions as well.

## After this PR

==COMMIT_MSG==
Include the maximum version information along with the minimum version.
==COMMIT_MSG==

## Possible downsides?
API breaks for ConfigureEndpointMinimumVersionsTask users.

### CODE REVIEW

**Comment by palantirtech** [comment]
Thanks for your interest in [palantir/gradle-conjure](https://github.com/palantir/gradle-conjure), @ylee088! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

File: gradle-conjure-api/src/main/java/com/palantir/gradle/conjure/api/EndpointVersion.java [hunk 1]
```diff
@@ -20,7 +20,7 @@
 import java.io.Serializable;
 import java.util.Objects;
 
-public final class EndpointMinimumVersion implements Serializable {
+public final class EndpointVersion implements Serializable {
```
Comment (esword) [inline gradle-conjure-api/src/main/java/com/palantir/gradle/conjure/api/EndpointVersion.java]: Thinking of naming - maybe EndpointVersionBounds?

File: gradle-conjure-api/src/main/java/com/palantir/gradle/conjure/api/EndpointVersion.java [hunk 2]
```diff
@@ -31,6 +31,9 @@
     @JsonProperty("min-version")
     private String minVersion;
 
+    @JsonProperty("max-version")
```
Comment (esword) [inline gradle-conjure-api/src/main/java/com/palantir/gradle/conjure/api/EndpointVersion.java]: Do you have tests for how reading in old data (without the max?) works?  Does this need to be marked as optional? Or have the build auto-insert "x.x.x" or something?

File: gradle-conjure-api/src/main/java/com/palantir/gradle/conjure/api/EndpointVersion.java (hunk 1 above)
Comment (ylee088) [reply gradle-conjure-api/src/main/java/com/palantir/gradle/conjure/api/EndpointVersion.java]: Done

File: gradle-conjure-api/src/main/java/com/palantir/gradle/conjure/api/EndpointVersion.java (hunk 2 above)
Comment (ylee088) [reply gradle-conjure-api/src/main/java/com/palantir/gradle/conjure/api/EndpointVersion.java]: Done, by assigning a default value as well as null-checking in the getter method.

**Review by esword — APPROVED** [review]

File: gradle-conjure-api/src/main/java/com/palantir/gradle/conjure/api/ConjureProductDependenciesExtension.java [hunk 3]
```diff
@@ -54,15 +54,15 @@ public final void serviceDependency(@DelegatesTo(ServiceDependency.class) Closur
         productDependencies.add(serviceDependency);
     }
 
-    public final void endpointVersion(@DelegatesTo(EndpointMinimumVersion.class) Closure<?> closure) {
+    public final void endpointVersion(@DelegatesTo(EndpointVersionBound.class) Closure<?> closure) {
         endpointVersions.add(providerFactory.provider(() -> {
-            EndpointMinimumVersion emv = new EndpointMinimumVersion();
+            EndpointVersionBound emv = new EndpointVersionBound();
```
Comment (fawind) [inline gradle-conjure-api/src/main/java/com/palantir/gradle/conjure/api/ConjureProductDependenciesExtension.java]: nit: Would rename `emv` to something that matches the new class name, e.g. `endpointVersionBound`.

File: gradle-conjure/src/main/java/com/palantir/gradle/conjure/ConfigureEndpointVersionBoundsTask.java [hunk 4]
```diff
@@ -58,21 +58,20 @@ final void action() {
...
+    private Manifest createManifest(Project project, List<EndpointVersionBound> versions) {
         org.gradle.api.plugins.JavaPluginConvention javaConvention =
                 project.getConvention().getPlugin(org.gradle.api.plugins.JavaPluginConvention.class);
         return javaConvention.manifest(manifest -> {
             String minVersionsString;
             try {
-                EndpointMinimumVersions emvs = EndpointMinimumVersions.builder()
-                        .minimumVersions(versions)
-                        .build();
+                EndpointVersionBounds emvs =
```
Comment (fawind) [inline gradle-conjure/src/main/java/com/palantir/gradle/conjure/ConfigureEndpointVersionBoundsTask.java]: nit: Same here regarding `emvs`

File: build.gradle [hunk 5]
```diff
@@ -57,7 +57,7 @@ subprojects {
     sourceCompatibility = 1.8
 
     tasks.withType(JavaCompile) {
-        options.compilerArgs += ['-Werror', '-Xlint:deprecation']
```
Comment (fawind) [inline build.gradle:60]: Why is this necessary? I would rather keep this and suppress individual occurrences if possible.

File: gradle-conjure-api/src/main/java/com/palantir/gradle/conjure/api/ConjureProductDependenciesExtension.java (hunk 3 above)
Comment (ylee088) [reply gradle-conjure-api/src/main/java/com/palantir/gradle/conjure/api/ConjureProductDependenciesExtension.java]: done

File: gradle-conjure/src/main/java/com/palantir/gradle/conjure/ConfigureEndpointVersionBoundsTask.java (hunk 4 above)
Comment (ylee088) [reply gradle-conjure/src/main/java/com/palantir/gradle/conjure/ConfigureEndpointVersionBoundsTask.java]: done

File: build.gradle (hunk 5 above)
Comment (ylee088) [reply build.gradle:60]: Few issues, e.g., EndpointMinimumVersions - @SuppressWarnings can't be used on import statements, and even if I fudge a little by using "import com.palantir.gradle.conjure.api.*;" (which checkstyle won't allow) or using a fully-qualified name in the code for the now-deprecated EndpointMinimumVersion, the Immutables will still generate a class with "import com.palantir.gradle.conjure.api.EndpointMinimumVersion;" which fails compilation due to -Werror and -xlint on deprecation. From what I can find, deprecation warnings can only be turned off globally on the compilation level, so I'd have to compile the generated classes separately.

I also looked into maybe using lint config that can exclude certain files, but I don't think that's available for JavaCompile.

**Review by fawind — APPROVED** [review]
